File: src/lib/weather.ts
import type { Checkpoint } from '../types'

const OPEN_METEO = 'https://api.open-meteo.com/v1/forecast'

export function weatherIcon(precipMm: number): string {
  return precipMm > 0 ? '🌧️' : '☀️'
}

interface HourlyBlock {
  hourly: {
    time: string[]
    temperature_2m: number[]
    precipitation: number[]
  }
}

const nearestHourIndex = (times: string[], targetISO: string): number => {
  const target = new Date(targetISO).getTime()
  let best = 0
  let bestDiff = Infinity
  for (let i = 0; i < times.length; i++) {
    // Open-Meteo hourly times are naive UTC ("2026-09-08T12:00") with timezone=UTC
    const diff = Math.abs(new Date(times[i] + 'Z').getTime() - target)
    if (diff < bestDiff) {
      bestDiff = diff
      best = i
    }
  }
  return best
}

/**
 * Fill in each checkpoint's weather from Open-Meteo in a single request (one
 * location per checkpoint, comma-joined). On any failure the checkpoints are
 * returned untouched so the schedule still renders.
 */
export async function attachWeather(
  checkpoints: Checkpoint[],
  fetchFn: typeof fetch = fetch,
): Promise<Checkpoint[]> {
  if (checkpoints.length === 0) return checkpoints

  const lat = checkpoints.map((c) => c.coord.lat.toFixed(4)).join(',')
  const lng = checkpoints.map((c) => c.coord.lng.toFixed(4)).join(',')
  const url =
    `${OPEN_METEO}?latitude=${lat}&longitude=${lng}` +
    `&hourly=temperature_2m,precipitation&timezone=UTC&forecast_days=3`

  let blocks: HourlyBlock[]
  try {
    const res = await fetchFn(url)
    if (!res.ok) return checkpoints
    const data = (await res.json()) as HourlyBlock | HourlyBlock[]
    blocks = Array.isArray(data) ? data : [data]
  } catch {
    return checkpoints
  }

  return checkpoints.map((cp, i) => {
    const block = blocks[i] ?? blocks[0]
    if (!block?.hourly?.time?.length) return cp
    const h = nearestHourIndex(block.hourly.time, cp.etaISO)
    const tempC = block.hourly.temperature_2m[h]
    const precipMm = block.hourly.precipitation[h]
    if (tempC == null || precipMm == null) return cp
    return { ...cp, weather: { tempC, precipMm, icon: weatherIcon(precipMm) } }
  })
}

File: cli/weather.py
"""Open-Meteo weather lookup — mirrors src/lib/weather.ts."""

from __future__ import annotations

from datetime import datetime, timezone

import requests

from .hazard import weather_icon
from .schedule import Checkpoint

OPEN_METEO = "https://api.open-meteo.com/v1/forecast"
TIMEOUT = 10


def _nearest_hour_index(times: list[str], target: datetime) -> int:
    target_utc = target.astimezone(timezone.utc).replace(tzinfo=None)
    best, best_diff = 0, None
    for i, t in enumerate(times):
        diff = abs((datetime.fromisoformat(t) - target_utc).total_seconds())
        if best_diff is None or diff < best_diff:
            best, best_diff = i, diff
    return best


def attach_weather(checkpoints: list[Checkpoint]) -> list[Checkpoint]:
    """Fill temp_c / precip_mm / icon on each checkpoint. On failure, leaves them unset."""
    if not checkpoints:
        return checkpoints
    lat = ",".join(f"{c['coord'][0]:.4f}" for c in checkpoints)
    lng = ",".join(f"{c['coord'][1]:.4f}" for c in checkpoints)
    try:
        resp = requests.get(
            OPEN_METEO,
            params={
                "latitude": lat,
                "longitude": lng,
                "hourly": "temperature_2m,precipitation",
                "timezone": "UTC",
                "forecast_days": 3,
            },
            timeout=TIMEOUT,
        )
        resp.raise_for_status()
        data = resp.json()
    except requests.RequestException:
        return checkpoints

    blocks = data if isinstance(data, list) else [data]
    for i, cp in enumerate(checkpoints):
        block = blocks[i] if i < len(blocks) else blocks[0]
        hourly = block.get("hourly") or {}
        times = hourly.get("time") or []
        if not times:
            continue
        h = _nearest_hour_index(times, cp["eta"])
        temp = hourly["temperature_2m"][h]
        precip = hourly["precipitation"][h]
        if temp is None or precip is None:
            continue
        cp["temp_c"] = temp
        cp["precip_mm"] = precip
        cp["icon"] = weather_icon(precip)
    return checkpoints
